src/main/java/net/neoforged/neoforge/registries/DeferredHolder.java

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.registries;
+
+import com.mojang.datafixers.util.Either;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import net.minecraft.core.Holder;
+import net.minecraft.core.HolderOwner;
+import net.minecraft.core.Registry;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.tags.TagKey;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A Deferred Holder is a Holder that is constructed with only a ResourceKey.<br> It will be populated with the underlying Holder from the registry when available.
+ *
+ * @param <T> The type of object being held by this DeferredHolder.
+ */
+public class DeferredHolder<R, T extends R> implements Holder<R>, Supplier<T> {
+    /**
+     * The resource key of the target object.
+     */
+    protected final ResourceKey<R> key;
+
+    /**
+     * The currently cached value.
+     */
+    @Nullable
+    private Holder<R> holder = null;
+
+    /**
+     * Creates a new DeferredHolder targeting the value with the specified name in the specified registry.
+     *
+     * @param <T>         The type of the target value.
+     * @param <R>         The registry type.
+     * @param registryKey The name of the registry the target value is a member of.
+     * @param valueName   The name of the target value.
+     */
+    public static <R, T extends R> DeferredHolder<R, T> create(ResourceKey<? extends Registry<R>> registryKey, ResourceLocation valueName) {
+        return create(ResourceKey.create(registryKey, valueName));
+    }
+
+    /**
+     * Creates a new DeferredHolder targeting the value with the specified name in the specified registry.
+     *
+     * @param <T>          The registry type.
+     * @param registryName The name of the registry the target value is a member of.
+     * @param valueName    The name of the target value.
+     */
+    public static <R, T extends R> DeferredHolder<R, T> create(ResourceLocation registryName, ResourceLocation valueName) {
+        return create(ResourceKey.createRegistryKey(registryName), valueName);
+    }
+
+    /**
+     * Creates a new DeferredHolder targeting the specified value.
+     *
+     * @param <T> The type of the target value.
+     * @param key The resource key of the target value.
+     */
+    public static <R, T extends R> DeferredHolder<R, T> create(ResourceKey<? super T> key) {
```

Comment:
I *think* ResourceKey is expected to be generified to `<R>`

Reply:
This can be validated by uses of ResourceKey in Holder, where `T` is the registry type: `public static <T> Holder.Reference<T> createStandAlone(HolderOwner<T> pOwner, ResourceKey<T> pKey)`

Similarly in `Registry` (where, once again, `T` is the registry type): `Optional<ResourceKey<T>> getResourceKey(T pValue);`

Reply:
Yeah you are right, `ResourceKey` should be `R` everywhere.